In [1]:
import pandas as pd 
import numpy as np 


In [2]:
#Load DataA.csv 

dataA = pd.read_csv('~/Documents/Uwaterloo_Study_Docs/657A_ECE/Assignments/ece_657a_assignments/Assignment_1/Datasets/DataA.csv')

In [3]:
print("dataA: \n", dataA)

dataA: 
        Unnamed: 0  fea.1  fea.2  fea.3  fea.4   fea.5  fea.6  fea.7  fea.8  \
0               1 -153.0  414.0  939.0 -161.0  1007.0   99.0 -210.0  948.0   
1               2 -150.0  420.0  939.0 -177.0  1008.0  103.0 -207.0  939.0   
2               3 -160.0  432.0  941.0 -162.0   982.0   98.0 -198.0  936.0   
3               4 -171.0  432.0  911.0 -174.0   999.0  115.0 -187.0  918.0   
4               5 -171.0    NaN  929.0 -189.0  1004.0  104.0 -198.0  939.0   
...           ...    ...    ...    ...    ...     ...    ...    ...    ...   
18995       18996    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   
18996       18997    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   
18997       18998    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   
18998       18999    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   
18999       19000    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   

       fea.9  ...  fea.72  fea.73  fea.74  fea.75  fea

In [4]:
# Display all columns
pd.set_option('display.max_columns', None)

## Shape of the Dataset

In [5]:
dataA.shape

(19000, 82)

#### The Dataset has 19000 samples and 82 features. However there is a Unamed column 'Unnamed' which is just the count which needs to be removed from the dataset 

## Checking all the Columns

In [6]:
dataA.columns

Index(['Unnamed: 0', 'fea.1', 'fea.2', 'fea.3', 'fea.4', 'fea.5', 'fea.6',
       'fea.7', 'fea.8', 'fea.9', 'fea.10', 'fea.11', 'fea.12', 'fea.13',
       'fea.14', 'fea.15', 'fea.16', 'fea.17', 'fea.18', 'fea.19', 'fea.20',
       'fea.21', 'fea.22', 'fea.23', 'fea.24', 'fea.25', 'fea.26', 'fea.27',
       'fea.28', 'fea.29', 'fea.30', 'fea.31', 'fea.32', 'fea.33', 'fea.34',
       'fea.35', 'fea.36', 'fea.37', 'fea.38', 'fea.39', 'fea.40', 'fea.41',
       'fea.42', 'fea.43', 'fea.44', 'fea.45', 'fea.46', 'fea.47', 'fea.48',
       'fea.49', 'fea.50', 'fea.51', 'fea.52', 'fea.53', 'fea.54', 'fea.55',
       'fea.56', 'fea.57', 'fea.58', 'fea.59', 'fea.60', 'fea.61', 'fea.62',
       'fea.63', 'fea.64', 'fea.65', 'fea.66', 'fea.67', 'fea.68', 'fea.69',
       'fea.70', 'fea.71', 'fea.72', 'fea.73', 'fea.74', 'fea.75', 'fea.76',
       'fea.77', 'fea.78', 'fea.79', 'fea.80', 'fea.81'],
      dtype='object')

#### Removing 'Unnamed' columns from the dataset  

In [7]:
dataA_updated = dataA.drop(['Unnamed: 0'], axis=1)
print("Updated Dataset after removal of 'Unnamed: 0' column: \n", dataA_updated.head(20))

Updated Dataset after removal of 'Unnamed: 0' column: 
     fea.1  fea.2  fea.3  fea.4   fea.5  fea.6  fea.7   fea.8  fea.9  fea.10  \
0  -153.0  414.0  939.0 -161.0  1007.0   99.0 -210.0   948.0  333.0   -19.0   
1  -150.0  420.0  939.0 -177.0  1008.0  103.0 -207.0   939.0  316.0     9.0   
2  -160.0  432.0  941.0 -162.0   982.0   98.0 -198.0   936.0  315.0   -10.0   
3  -171.0  432.0  911.0 -174.0   999.0  115.0 -187.0   918.0  338.0    34.0   
4  -171.0    NaN  929.0 -189.0  1004.0  104.0 -198.0   939.0  350.0    60.0   
5  -171.0  432.0  924.0 -179.0  1011.0   85.0 -204.0   945.0  336.0    94.0   
6  -169.0  429.0  949.0 -175.0  1007.0  102.0 -188.0   914.0  322.0   154.0   
7  -160.0  423.0  927.0 -195.0   996.0  123.0 -213.0   925.0  302.0   128.0   
8  -163.0  432.0  929.0 -178.0   994.0  101.0 -186.0   946.0  296.0   166.0   
9  -156.0  415.0  936.0 -186.0  1014.0  111.0 -195.0   960.0  280.0   202.0   
10 -153.0  413.0  923.0 -187.0   993.0   91.0 -193.0   970.0  282.0   233.0

In [8]:
dataA_updated.shape

(19000, 81)

#### The Column 'Unnamed: 0' has been removed from the dataset as we can see in the above data and now the dimension of the dataset has become (19000, 81) which means it has 19000 samples and 81 features in total

## Identifying Missing Values in the Dataset 


In [9]:
dataA_updated.isnull()

fea.1  fea.2  fea.3  fea.4  fea.5  fea.6  fea.7  fea.8  fea.9  fea.10  \
0      False  False  False  False  False  False  False  False  False   False   
1      False  False  False  False  False  False  False  False  False   False   
2      False  False  False  False  False  False  False  False  False   False   
3      False  False  False  False  False  False  False  False  False   False   
4      False   True  False  False  False  False  False  False  False   False   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...     ...   
18995   True   True   True   True   True   True   True   True   True    True   
18996   True   True   True   True   True   True   True   True   True    True   
18997   True   True   True   True   True   True   True   True   True    True   
18998   True   True   True   True   True   True   True   True   True    True   
18999   True   True   True   True   True   True   True   True   True    True   

       fea.11  fea.12  fea.13  fea.14  fea.15  fea.16  fea.17  fea.18  fea.19  \
0       False   False   False   False   False   False   False   False   False   
1       False   False   False   False   False   False   False   False   False   
2       False   False   False   False   False   False   False   False   False   
3       False   False   False   False   False   False   False   False   False   
4       False   False   False   False   False   False   False   False   False   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995    True    True    True    True    True    True    True    True    True   
18996    True    True    True    True    True    True    True    True    True   
18997    True    True    True    True    True    True    True    True    True   
18998    True    True    True    True    True    True    True    True    True   
18999    True    True    True    True    True    True    True    True    True   

       fea.20  fea.21  fea.22  fea.23  fea.24  fea.25  fea.26  fea.27  fea.28  \
0       False   False   False   False   False   False   False   False   False   
1       False   False   False   False   False   False   False   False   False   
2       False   False   False   False   False   False   False   False   False   
3       False   False   False   False   False   False   False   False   False   
4       False   False   False   False   False   False   False   False   False   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995    True    True    True    True    True    True    True    True    True   
18996    True    True    True    True    True    True    True    True    True   
18997    True    True    True    True    True    True    True    True    True   
18998    True    True    True    True    True    True    True    True    True   
18999    True    True    True    True    True    True    True    True    True   

       fea.29  fea.30  fea.31  fea.32  fea.33  fea.34  fea.35  fea.36  fea.37  \
0       False   False   False   False   False    True    True    True   False   
1       False   False   False   False   False    True    True    True   False   
2       False   False   False   False   False    True    True    True   False   
3       False   False   False   False   False    True    True    True   False   
4       False   False   False   False   False    True    True    True   False   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995    True    True    True    True    True    True    True    True    True   
18996    True    True    True    True    True    True    True    True    True   
18997    True    True    True    True    True    True    True    True    True   
18998    True    True    True    True    True    True    True    True    True   
18999    True    True    True    True    True    True    True    True    True   

       fea.38  fea.39  fea.40  fea.41  fea.42  fea.43  fea.44  fea.45  fea.46  \
0       False   False   False   False   False 

#### Here We can see in the dataset there are many missing value in the Dataset 

## Number of Missing values per columns

In [10]:
# Find the Total number of missing values in each column
dataA_updated.isnull().sum()

fea.1     1187
fea.2     1188
fea.3     1187
fea.4      800
fea.5      800
          ... 
fea.77     773
fea.78     773
fea.79     773
fea.80     773
fea.81     773
Length: 81, dtype: int64

#### As we can see from the above table that the dataset contains too many missing values and each feature has different number of missing values. As from the table it is clear that feature.1 has 1187 missing values, feature.2 has 1188 missing values while from feature.37 to feature.81 all have a missing value of 773 which is the lowest value missing among all the features. Whereas feature.34 to feature.36 have 18999 missing values which highest among all the features in the entire dataset. 

## Total number of Missing values in the dataset

In [11]:
dataA_updated.isnull().sum().sum()

124053

#### In total there are 124053 missing values in the dataset

## Statistical Description of the Dataset

In [12]:
dataA_updated.describe()

fea.1         fea.2         fea.3         fea.4         fea.5  \
count  17813.000000  17812.000000  17813.000000  18200.000000  18200.000000   
mean    -132.812384    698.264485    597.541402   -307.128462    909.548077   
std      284.183187    375.672475    396.654659    183.151634    193.963300   
min    -2724.000000   -855.000000  -2196.000000  -1365.000000   -245.000000   
25%     -179.000000    360.000000    304.000000   -409.000000    860.000000   
50%     -100.000000    811.000000    574.000000   -266.000000    969.500000   
75%      -15.000000    984.000000    955.000000   -167.000000   1006.000000   
max     1887.000000   2531.000000   2941.000000    609.000000   1833.000000   

              fea.6         fea.7         fea.8         fea.9        fea.10  \
count  18200.000000  18099.000000  18099.000000  18099.000000  18043.000000   
mean     -32.760824     61.974363    899.313498     81.650478    356.638752   
std      254.001018    317.393784    196.829252    327.904371    343.131382   
min     -920.000000  -1580.000000   -149.000000  -1624.000000  -1792.000000   
25%     -144.000000   -131.000000    854.000000   -155.000000    158.000000   
50%      -39.000000     70.000000    946.000000     41.000000    380.000000   
75%       45.000000    251.000000    997.000000    315.000000    583.000000   
max     1215.000000   1490.000000   1682.000000   1096.000000   2202.000000   

             fea.11        fea.12        fea.13        fea.14        fea.15  \
count  18044.000000  18044.000000  17950.000000  17950.000000  17950.000000   
mean     330.971569    584.833130    583.589861    421.624290    -38.719833   
std      543.517364    282.078053    307.610433    543.293018    495.913749   
min    -1545.000000  -1079.000000  -1710.000000  -1120.000000  -1634.000000   
25%     -142.000000    370.000000    347.000000   -131.000000   -484.750000   
50%      395.000000    610.000000    628.000000    524.000000    -14.000000   
75%      786.000000    814.000000    818.000000    886.000000    274.000000   
max     2047.000000   2152.000000   2408.000000   2212.000000   2620.000000   

             fea.16        fea.17        fea.18        fea.19        fea.20  \
count  18202.000000  18202.000000  18202.000000  17964.000000  17964.000000   
mean      71.807768    780.358257   -461.605043    274.476843    893.086451   
std      258.937246    291.642323    261.243774    298.191766    216.179301   
min    -1089.000000   -289.000000  -1379.000000  -2269.000000   -786.000000   
25%        6.000000    696.000000   -649.000000    252.000000    878.000000   
50%       94.000000    905.000000   -398.000000    302.000000    921.000000   
75%      180.000000    967.000000   -282.000000    356.000000    945.000000   
max      878.000000   1653.000000   1292.000000   2335.000000   2466.000000   

             fea.21        fea.22        fea.23        fea.24        fea.25  \
count  17964.000000  17677.000000  17677.000000  17677.000000  18106.000000   
mean     -89.299933   -622.139334    233.764044   -131.400068   -360.661825   
std      221.071614    359.196753    587.480524    460.080469    223.329186   
min    -1944.000000  -3172.000000  -2324.000000  -2734.000000  -1873.000000   
25%     -192.000000   -876.000000   -193.000000   -468.000000   -475.000000   
50%      -70.000000   -655.000000    267.000000   -136.000000   -342.000000   
75%       31.000000   -371.000000    731.000000    184.000000   -205.000000   
max     1968.000000   2229.000000   2677.000000   2401.000000    711.000000   

             fea.26        fea.27        fea.28        fea.29        fea.30  \
count  18106.000000  18106.000000  18106.000000  18106.000000  18106.000000   
mean     875.943278    269.614934   -570.500166    432.334420    427.547277   
std      253.761954    371.407492    307.329729    497.493607    204.477583   
min     -442.000000  -1430.000000  -1933.000000  -1103.000000   -765.000000   
25%      770.000000     11.000000   -825.000000    -59.0000

#### As we can see from the above statistical description of the dataset it clear that the rows of feature.34, feature.35 and feature.36 is entirely missing so there is no statistical value for those columns. We have to remove the entire feature.34, feature.35 and feature.36 from the dataset.

## Drop feature.34, feature.35 and feature.36

In [13]:
dataA_updated = dataA_updated.drop(['fea.34', 'fea.35', 'fea.36'], axis=1)

In [14]:
dataA_updated

fea.1  fea.2  fea.3  fea.4   fea.5  fea.6  fea.7  fea.8  fea.9  fea.10  \
0     -153.0  414.0  939.0 -161.0  1007.0   99.0 -210.0  948.0  333.0   -19.0   
1     -150.0  420.0  939.0 -177.0  1008.0  103.0 -207.0  939.0  316.0     9.0   
2     -160.0  432.0  941.0 -162.0   982.0   98.0 -198.0  936.0  315.0   -10.0   
3     -171.0  432.0  911.0 -174.0   999.0  115.0 -187.0  918.0  338.0    34.0   
4     -171.0    NaN  929.0 -189.0  1004.0  104.0 -198.0  939.0  350.0    60.0   
...      ...    ...    ...    ...     ...    ...    ...    ...    ...     ...   
18995    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   
18996    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   
18997    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   
18998    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   
18999    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   

       fea.11  fea.12  fea.13  fea.14  fea.15  fea.16  fea.17  fea.18  fea.19  \
0      -587.0   810.0   902.0  -140.0  -468.0   -28.0  1016.0  -191.0   358.0   
1      -605.0   835.0   897.0  -136.0  -473.0   -38.0   989.0  -160.0   346.0   
2      -580.0   802.0   902.0  -139.0  -454.0   -19.0   992.0  -170.0   345.0   
3      -579.0   849.0   910.0  -151.0  -460.0   -18.0  1015.0  -185.0   345.0   
4      -584.0   843.0   917.0  -151.0  -463.0   -12.0   996.0  -209.0   345.0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18996     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18997     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18998     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18999     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

       fea.20  fea.21  fea.22  fea.23  fea.24  fea.25  fea.26  fea.27  fea.28  \
0       895.0  -189.0  -511.0  -782.0  -212.0  -241.0   472.0   938.0  -882.0   
1       889.0  -215.0  -609.0  -773.0  -202.0  -255.0   509.0   934.0  -878.0   
2       898.0  -200.0  -606.0  -753.0  -116.0  -272.0   492.0   919.0  -864.0   
3       888.0  -220.0  -638.0  -718.0  -354.0  -275.0   517.0   919.0  -847.0   
4       886.0  -204.0  -723.0  -679.0  -220.0  -271.0   525.0   923.0  -862.0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18996     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18997     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18998     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18999     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

       fea.29  fea.30  fea.31  fea.32  fea.33  fea.37  fea.38  fea.39  fea.40  \
0       -61.0   543.0   141.0  -150.0   965.0  -989.0  -158.0    41.0    78.0   
1       -50.0   515.0   130.0  -179.0   995.0  -991.0  -157.0    43.0    48.0   
2       -56.0   540.0   122.0  -162.0   946.0  -988.0  -171.0    41.0    68.0   
3       -66.0   546.0   145.0  -144.0   939.0  -994.0  -159.0    42.0    32.0   
4       -70.0   531.0   153.0  -167.0   970.0 -1002.0  -161.0    42.0    58.0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18996     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18997     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18998     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
18999     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

       fea.41  fea.42  fea.43  fea.44  fea.45  fea.46  fea.47  fea.48  fea.49  \
0       -49.0   -13.0   643.0   10

#### Feature.34, Feature.35 and feature.36 has been dropped from the dataset

In [15]:
dataA_updated.isnull()

fea.1  fea.2  fea.3  fea.4  fea.5  fea.6  fea.7  fea.8  fea.9  fea.10  \
0      False  False  False  False  False  False  False  False  False   False   
1      False  False  False  False  False  False  False  False  False   False   
2      False  False  False  False  False  False  False  False  False   False   
3      False  False  False  False  False  False  False  False  False   False   
4      False   True  False  False  False  False  False  False  False   False   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...     ...   
18995   True   True   True   True   True   True   True   True   True    True   
18996   True   True   True   True   True   True   True   True   True    True   
18997   True   True   True   True   True   True   True   True   True    True   
18998   True   True   True   True   True   True   True   True   True    True   
18999   True   True   True   True   True   True   True   True   True    True   

       fea.11  fea.12  fea.13  fea.14  fea.15  fea.16  fea.17  fea.18  fea.19  \
0       False   False   False   False   False   False   False   False   False   
1       False   False   False   False   False   False   False   False   False   
2       False   False   False   False   False   False   False   False   False   
3       False   False   False   False   False   False   False   False   False   
4       False   False   False   False   False   False   False   False   False   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995    True    True    True    True    True    True    True    True    True   
18996    True    True    True    True    True    True    True    True    True   
18997    True    True    True    True    True    True    True    True    True   
18998    True    True    True    True    True    True    True    True    True   
18999    True    True    True    True    True    True    True    True    True   

       fea.20  fea.21  fea.22  fea.23  fea.24  fea.25  fea.26  fea.27  fea.28  \
0       False   False   False   False   False   False   False   False   False   
1       False   False   False   False   False   False   False   False   False   
2       False   False   False   False   False   False   False   False   False   
3       False   False   False   False   False   False   False   False   False   
4       False   False   False   False   False   False   False   False   False   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995    True    True    True    True    True    True    True    True    True   
18996    True    True    True    True    True    True    True    True    True   
18997    True    True    True    True    True    True    True    True    True   
18998    True    True    True    True    True    True    True    True    True   
18999    True    True    True    True    True    True    True    True    True   

       fea.29  fea.30  fea.31  fea.32  fea.33  fea.37  fea.38  fea.39  fea.40  \
0       False   False   False   False   False   False   False   False   False   
1       False   False   False   False   False   False   False   False   False   
2       False   False   False   False   False   False   False   False   False   
3       False   False   False   False   False   False   False   False   False   
4       False   False   False   False   False   False   False   False   False   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
18995    True    True    True    True    True    True    True    True    True   
18996    True    True    True    True    True    True    True    True    True   
18997    True    True    True    True    True    True    True    True    True   
18998    True    True    True    True    True    True    True    True    True   
18999    True    True    True    True    True    True    True    True    True   

       fea.41  fea.42  fea.43  fea.44  fea.45  fea.46  fea.47  fea.48  fea.49  \
0       False   False   False   False   False 

In [16]:
dataA_updated.isnull().sum()

fea.1     1187
fea.2     1188
fea.3     1187
fea.4      800
fea.5      800
          ... 
fea.77     773
fea.78     773
fea.79     773
fea.80     773
fea.81     773
Length: 78, dtype: int64

#### The dataset still contains missing values in the other Feature sets. So we will apply Imputation of missing values learned during lectures. 

# Imputing Missing Values
### 1. Replace missing data by Mean values for each columns

In [17]:
dataA_updated.fillna(dataA_updated.mean(), inplace=True)

In [18]:
dataA_updated.describe()

fea.1         fea.2         fea.3         fea.4         fea.5  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean    -132.812384    698.264485    597.541402   -307.128462    909.548077   
std      275.162567    363.737566    384.063939    179.254138    189.835729   
min    -2724.000000   -855.000000  -2196.000000  -1365.000000   -245.000000   
25%     -172.000000    366.000000    322.750000   -402.000000    868.000000   
50%     -110.000000    756.000000    597.541402   -277.000000    964.000000   
75%      -20.000000    977.000000    950.000000   -169.000000   1005.000000   
max     1887.000000   2531.000000   2941.000000    609.000000   1833.000000   

              fea.6         fea.7         fea.8         fea.9        fea.10  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean     -32.760824     61.974363    899.313498     81.650478    356.638752   
std      248.595835    309.776406    192.105395    320.034742    334.377778   
min     -920.000000  -1580.000000   -149.000000  -1624.000000  -1792.000000   
25%     -139.000000   -123.000000    861.000000   -143.000000    170.000000   
50%      -32.760824     61.974363    940.000000     68.000000    359.000000   
75%       40.000000    239.000000    994.000000    303.000000    570.000000   
max     1215.000000   1490.000000   1682.000000   1096.000000   2202.000000   

             fea.11        fea.12        fea.13        fea.14        fea.15  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean     330.971569    584.833130    583.589861    421.624290    -38.719833   
std      529.666403    274.889595    298.989418    528.066821    482.015391   
min    -1545.000000  -1079.000000  -1710.000000  -1120.000000  -1634.000000   
25%     -125.000000    382.000000    361.000000   -110.000000   -456.000000   
50%      337.000000    594.000000    594.000000    481.000000    -38.719833   
75%      770.000000    804.000000    809.000000    863.000000    245.000000   
max     2047.000000   2152.000000   2408.000000   2212.000000   2620.000000   

             fea.16        fea.17        fea.18        fea.19        fea.20  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean      71.807768    780.358257   -461.605043    274.476843    893.086451   
std      253.440945    285.451810    255.698513    289.947729    210.202643   
min    -1089.000000   -289.000000  -1379.000000  -2269.000000   -786.000000   
25%       11.000000    708.000000   -637.000000    259.000000    881.000000   
50%       86.000000    895.000000   -412.000000    299.000000    917.000000   
75%      174.000000    964.000000   -287.000000    352.000000    943.000000   
max      878.000000   1653.000000   1292.000000   2335.000000   2466.000000   

             fea.21        fea.22        fea.23        fea.24        fea.25  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean     -89.299933   -622.139334    233.764044   -131.400068   -360.661825   
std      214.959699    346.464727    566.656792    443.772537    218.011487   
min    -1944.000000  -3172.000000  -2324.000000  -2734.000000  -1873.000000   
25%     -186.000000   -859.000000   -156.000000   -444.000000   -467.000000   
50%      -81.000000   -622.139334    233.764044   -131.400068   -353.000000   
75%       28.000000   -393.000000    695.000000    156.000000   -212.000000   
max     1968.000000   2229.000000   2677.000000   2401.000000    711.000000   

             fea.26        fea.27        fea.28        fea.29        fea.30  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean     875.943278    269.614934   -570.500166    432.334420    427.547277   
std      247.719619    362.563895    300.011889    485.647768    199.608760   
min     -442.000000  -1430.000000  -1933.000000  -1103.000000   -765.000000   
25%      781.000000     24.000000   -819.000000    -54.0000

In [18]:
# Check for number of missing values in the entire dataset
dataA_updated.isnull().sum().sum()

0

#### Using Mean Imputation we have replaced all the missing values in the dataset as we can get from the above code